python/lsst/ip/diffim/zogy.py

```diff
@@ -0,0 +1,839 @@
+
+        return pipeBase.Struct(D=D, D_var=D_var)
+
+    def _doConvolve(self, exposure, kernel, recenterKernel=False):
+        """! Convolve an Exposure with a decorrelation convolution kernel.
+
+        Parameters
+        ----------
+        exposure : lsst.afw.image.Exposure to be convolved.
+        kernel : 2D numpy.array to convolve the image with
+
+        Returns
+        -------
+        A new lsst.afw.image.Exposure with the convolved pixels and the (possibly
+        re-centered) kernel.
+
+        Notes
+        -----
+        - We use afwMath.convolve() but keep scipy.convolve for debugging.
+        - We optionally re-center the kernel if necessary and return the possibly
+          re-centered kernel
+        """
+        kernelImg = afwImage.ImageD(kernel.shape[0], kernel.shape[1])
+        kernelImg.getArray()[:, :] = kernel
+        kern = afwMath.FixedKernel(kernelImg)
+        if recenterKernel:
+            maxloc = np.unravel_index(np.argmax(kernel), kernel.shape)
+            kern.setCtrX(maxloc[0])
+            kern.setCtrY(maxloc[1])
+        outExp = exposure.clone()  # Do this to keep WCS, PSF, masks, etc.
+        convCntrl = afwMath.ConvolutionControl(doNormalize=False, doCopyEdge=False,
+                                               maxInterpolationDistance=0)
+        try:
+            afwMath.convolve(outExp.getMaskedImage(), exposure.getMaskedImage(), kern, convCntrl)
+        except:  # Allow exposure to actually be an image/maskedImage
+            afwMath.convolve(outExp, exposure, kern, convCntrl)
+
+        return outExp, kern
+
+    # In all functions, im1 is R (reference, or template) and im2 is N (new, or science)
+    def computeDiffimImageSpace(self, padSize=None, debug=False, **kwargs):
+        """Compute ZOGY diffim `D` using image-space convlutions
+
+        This method is still being debugged as it results in artifacts
+        when the PSFs are noisy (see module-level docstring). Thus
+        there are several options still enabled by the `debug` flag,
+        which are disabled by defult.
+
+        Parameters
+        ----------
+        padSize : int, the amount to pad the PSFs by
+        debug : bool, flag to enable debugging tests and options
+
+        Returns
+        -------
+        D : lsst.afw.Exposure
+           the proper image difference, including correct variance,
+           masks, and PSF
+        """
+        preqs = self.computePrereqs(padSize=padSize)
+
+        delta = 0.  # 1
+        Kr_hat = (preqs.Pr_hat + delta) / (preqs.denom + delta)
+        Kn_hat = (preqs.Pn_hat + delta) / (preqs.denom + delta)
+        Kr = np.fft.ifft2(Kr_hat).real
+        Kr = np.roll(np.roll(Kr, -1, 0), -1, 1)
+        Kn = np.fft.ifft2(Kn_hat).real
+        Kn = np.roll(np.roll(Kn, -1, 0), -1, 1)
+
+        # Using this block makes it slightly faster, but ~25% worse artifacts:
+        padSize = self.padSize if padSize is None else padSize
+        if debug and padSize > 0:
```

Comment:
If you're referring to the "if debug..." block, this doesn't set self.padSize to zero. Neither does the line above it (if that's the one you're actually talking about). 

But as you suggest below I will move it and the next block to a "filterPsf" function.